src/dex.hs

```diff
@@ -37,6 +40,29 @@ data EvalMode = ReplMode String
 
 data CmdOpts = CmdOpts EvalMode (Maybe FilePath) EvalConfig
 
+
+
+filteredCompletions candidates str = map simpleCompletion $
+  filter (str `isPrefixOf`) candidates
+
+completions :: CompletionFunc (StateT TopEnv IO)
+completions (line, _) = do
+  env <- get
+  let varNames = map (show . pretty) $ envNames env
+  -- note: line and thus word and rest have character order reversed
+  let (word, rest) = break (== ' ') line
+  let anywhereKeywords = ["def", "for", "rof", "case", "data", "where", "of", "if", "then", "else", "interface", "instance", "do", "view"]
+  let startoflineKeywords = ["%bench \"", ":p", ":t", ":html", ":export"]
+  let candidates = varNames ++ anywhereKeywords ++ if null rest
+                                                    then startoflineKeywords
+                                                    else []
+  return (rest, filteredCompletions candidates $ reverse word)
+
+dexCompletions = completeQuotedWord (Just '\\') "\"'" listFiles completions
+
+hasklineSettings :: Settings (StateT TopEnv IO)
+hasklineSettings = setComplete dexCompletions defaultSettings
```

Comment:
nit: small local definitions like this one (or `dexCompletions`) are usually better placed in the `where` clauses of the functions that use them (in this case`runMode`).

Reply:
It seems like if i do that it is too far away from where it is used.
So I have put it side that `ReplMode` case